Source/Core/VideoCommon/NetplayUI.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "VideoCommon/NetplayUI.h"
+
+#include <utility>
+#include <vector>
+
+#include "imgui.h"
+
+constexpr size_t MAX_BACKLOG_SIZE = 100;
+
+static char message_buf[160] = {};
+static bool visible = false;
+static std::vector<std::pair<std::string, ImVec4>> messages;
+static std::function<void(const std::string)> message_callback;
+
+void NetplayUI::Display()
+{
+  if (!visible)
+    return;
+
+  const float scale = ImGui::GetIO().DisplayFramebufferScale.x;
+
+  ImGui::SetNextWindowPos(ImVec2(10.0f * scale, 10.0f * scale), ImGuiCond_FirstUseEver);
+  ImGui::SetNextWindowSizeConstraints(ImVec2(200.0f * scale, 400.0f * scale),
+                                      ImGui::GetIO().DisplaySize);
+  if (!ImGui::Begin("Chat", nullptr, ImGuiWindowFlags_None))
+  {
+    ImGui::End();
+    return;
+  }
+
+  ImGui::BeginChild("Scrolling", ImVec2(220 * scale, -65), true,
+                    ImGuiWindowFlags_HorizontalScrollbar);
+  for (const auto& msg : messages)
+  {
+    ImGui::TextColored(msg.second, "%s", msg.first.c_str());
+  }
+  ImGui::EndChild();
+
+  ImGui::Spacing();
+
+  ImGui::InputText("", message_buf, IM_ARRAYSIZE(message_buf));
+
+  ImGui::SameLine();
+
+  if (ImGui::Button("Send"))
+  {
+    if (message_callback != nullptr)
+      message_callback(message_buf);
+
+    // 'Empty' the buffer
+    message_buf[0] = '\0';
+  }
+
+  ImGui::End();
+}
+
+void NetplayUI::SetVisible(bool v)
+{
+  visible = v;
+}
+
+void NetplayUI::AppendChat(const std::string message, const std::string color)
+{
+  if (messages.size() > MAX_BACKLOG_SIZE)
+    messages.erase(messages.begin());
+
+  ImVec4 c;
+
+  if (color == "magenta")
+    c = {1.0f, 0.0f, 1.0f, 1.0f};
+  else if (color == "red")
+    c = {1.0f, 0.0f, 0.0f, 1.0f};
+  else if (color == "green")
+    c = {0.0f, 1.0f, 0.0f, 1.0f};
+  else if (color == "yellow")
+    c = {1.0f, 1.0f, 0.0f, 1.0f};
+  else
+    c = {1.0f, 1.0f, 1.0f, 1.0f};
+
+  messages.push_back({message, c});
+}
+
+void NetplayUI::SetSendChatMessageCallback(std::function<void(const std::string)> callback)
+{
+  message_callback = callback;
```

Comment:
```suggestion
  message_callback = std::move(callback);
```

std::function is allowed to heap allocate